org.eclipse.xtext.tests/src/org/eclipse/xtext/enumrules/ResourceStorageTest.java

```diff
@@ -0,0 +1,168 @@
+			return models.get(uri);
+		}
+	}
+	
+	private ParserTestHelper helper;
+
+	@Inject
+	private ResourceStorageFacade resourceStorageFacade;
+	
+	@Override
+	public void setUp() throws Exception {
+		super.setUp();
+		with(CustomSetup.class);
+		helper = new ParserTestHelper(getResourceFactory(), getParser(), get(Keys.RESOURCE_SET_KEY),getCurrentFileExtension());
+		injectMembers(this);
+	}
+
+	@Test
+	public void testWriteAndLoad() throws Exception {
+		String modelAsString = "kindOfKeyword Hoo reference Hoo";
+		
+		StorageAwareResource originalResource = parse(modelAsString);
+		DetachableParseResult originalParseResult = (DetachableParseResult) originalResource.getParseResult();
+		RootNode originalRootNode = (RootNode) originalParseResult.getRootNode();
+		
+		EntityWithEnumAndReference model = (EntityWithEnumAndReference) originalResource.getContents().get(0);
+		assertEquals("Hoo", model.getName());
+		EntityWithEnumAndReference proxy = (EntityWithEnumAndReference) model.eGet(EnumAndReferenceTestLanguagePackage.Literals.ENTITY_WITH_ENUM_AND_REFERENCE__REF, false);
+		assertTrue(proxy.eIsProxy());
+		
+		EcoreUtil.resolveAll(originalResource);
+		assertTrue(originalResource.getErrors().isEmpty());
+		
+		ByteArrayOutputStream bout = new ByteArrayOutputStream();
+		resourceStorageFacade.setStoreNodeModel(true);
+		resourceStorageFacade.createResourceStorageWritable(bout).writeResource(originalResource);
+		ResourceStorageLoadable in = resourceStorageFacade.createResourceStorageLoadable(new ByteArrayInputStream(bout.toByteArray()));
+		StorageAwareResource reloadedResource = (StorageAwareResource) originalResource.getResourceSet().createResource(URI.createURI("synthetic:/Test." + originalResource.getURI().fileExtension()));
+		InMemoryURIConverter converter = new InMemoryURIConverter();
+		converter.addModel(reloadedResource.getURI().toString(), modelAsString);
+		reloadedResource.getResourceSet().setURIConverter(converter);
+		originalResource.getResourceSet().getResources().add(reloadedResource);
+		reloadedResource.loadFromStorage(in);
+		EntityWithEnumAndReference reloadedModel = (EntityWithEnumAndReference) reloadedResource.getContents().get(0);
+		EqualityHelper equalityHelper = new EqualityHelper();
+		assertTrue(equalityHelper.equals(model, reloadedModel));
+		
+		DetachableParseResult reloadedParseResult = (DetachableParseResult) reloadedResource.getParseResult();
+		RootNode reloadedRootNode = (RootNode) reloadedParseResult.getRootNode();
+		
+		assertEqualNodes(originalRootNode, reloadedRootNode, equalityHelper);
+		
+		assertSame(reloadedModel, reloadedModel.getRef());
+	}
+	
+	protected void assertEqualNodes(RootNode expected, RootNode actual, EqualityHelper equalityHelper) throws IOException {
+		Iterator<INode> expectedIter = expected.getAsTreeIterable().iterator();
+		Iterator<INode> actualIter = actual.getAsTreeIterable().iterator();
+		while(expectedIter.hasNext()) {
+			assertTrue(actualIter.hasNext());
```

Comment:
Just my preference, so please feel free to ignore that, but I prefer to specify a message string when using assertTrue for better understanding in case the test fails

Reply:
You're right about this. I expect that I'll do more refinements to the code and add the assertions messages in a subsequent step.